ocaml/idl/json_backend/gen_json.ml

```diff
@@ -541,31 +541,71 @@ let write_to_dir dir maybe_file =
   Option.iter write maybe_file
 
 module Version = struct
-  type t = int list
-
-  let rec compare x y =
-    match (x, y) with
-    | x :: xs, y :: ys when x = y ->
-        compare xs ys
-    | x :: _, y :: _ ->
-        Int.compare x y
-    | x :: _, [] when x = 0 ->
-        0
-    | _ :: _, [] ->
-        1
-    | [], y :: _ when y = 0 ->
-        0
-    | [], _ :: _ ->
-        -1
-    | [], [] ->
-        0
+  (* versions ending in -next are considered unreleased *)
+  type t = int list * string option
+
+  (* orders like this:
+     - 22.10.0
+     - 22.11.0
+     - 22.11.0-next
+     - 22.12.0
+     From the way the lifecycle binary generates versions there only will be a
+     single -next version, with the highest version number
+  *)
+  let compare (xs, x_next) (ys, y_next) =
+    let rec compare_list x y =
+      match (x, y) with
+      | x :: xs, y :: ys when x = y ->
+          compare_list xs ys
```

Comment:
4.12 has List.compare: https://v2.ocaml.org/api/List.html so this could be written more simply